universe/Condition.cpp

```diff
@@ -9956,6 +9965,249 @@ unsigned int Not::GetCheckSum() const {
+// OrderedAlternativesOf
+///////////////////////////////////////////////////////////
+void FCMoveContent(Condition::ObjectSet& from_set, Condition::ObjectSet& to_set) {
+    to_set.insert(to_set.end(),
+                  std::make_move_iterator(from_set.begin()),
+                  std::make_move_iterator(from_set.end()));
+    from_set.clear();
+}
+
+OrderedAlternativesOf::OrderedAlternativesOf(
+    std::vector<std::unique_ptr<ConditionBase>>&& operands) :
+    ConditionBase(),
+    m_operands(std::move(operands))
+{}
+
+bool OrderedAlternativesOf::operator==(const ConditionBase& rhs) const {
+    if (this == &rhs)
+        return true;
+    if (typeid(*this) != typeid(rhs))
+        return false;
+
+    const OrderedAlternativesOf& rhs_ = static_cast<const OrderedAlternativesOf&>(rhs);
+
+    if (m_operands.size() != rhs_.m_operands.size())
+        return false;
+    for (unsigned int i = 0; i < m_operands.size(); ++i) {
+        CHECK_COND_VREF_MEMBER(m_operands.at(i))
+    }
+
+    return true;
+}
+
+void OrderedAlternativesOf::Eval(const ScriptingContext& parent_context,
+                                 ObjectSet& matches, ObjectSet& non_matches,
+                                 SearchDomain search_domain/* = NON_MATCHES*/) const
+{
+    std::shared_ptr<const UniverseObject> no_object;
+    ScriptingContext local_context(parent_context, no_object);
+
+    if (m_operands.empty()) {
+        ErrorLogger() << "OrderedAlternativesOf::Eval given no operands!";
+        return;
+    }
+    for (auto& operand : m_operands) {
+        if (!operand) {
+            ErrorLogger() << "OrderedAlternativesOf::Eval given null operand!";
+            return;
+        }
+    }
+
+    // OrderedAlternativesOf [ A B C ] matches all candidates which match the topmost condition.
+    // If any candidate matches A, then all candidates that match A are matched,
+    // or if no candidate matches A but any candidate matches B, then all candidates that match B are matched,
+    // or if no candidate matches A or B but any candidate matches C, then all candidates that match C are matched.
+    // If no candidate matches A, B, or C, then nothing is matched.
+    //
+    // Not OrderedAlternativesOf [ A B C ] finds the topmost condition which has matches and then matches its non-matches.
+    // If any candidate matches A, then all candidates that do not match A are matched,
+    // or if no candidates match A but any candidate matches B, then all candidates that do not match B are matched,
+    // or if no candidate matches A or B but any candidate matches C, then all candidates that do not match C are matched.
+    // If no candidate matches A, B, or C, then all candidates are matched.
+    if (search_domain == NON_MATCHES) {
+        // check all operand conditions on all objects in the non matches set
+        // stop if the operand condition has at least one match
+        // Return non-matches if there are any
```

Comment:
// check each operand condition on objects in the input matches and non_matches sets, until an operand condition matches an object. if an operand condition is selected, apply it to the input non_matches set, moving matching candidates to matches. if no operand condition is selected, because no candidate is matched by any operand condition, then do nothing.

Reply:
Do you propose a change in semantics here? Probably my current implementation is actually wrong.

Because at the moment it is: if no operand condition is selected, because no candidate is matched by any operand condition, then move all objects from the matches input set into the non_matches set.

I will test using an OR condition.

Reply:
Not sure if it is a change. I was suggesting more details for the comment about how I thought / expected it did or should work.

Reply:
yes, confirmed there was a bug, fixed